Add missing `Type Hints` to torch. issue #7318

kimdwkimdw opened the pull request:
https://github.com/pytorch/pytorch/issues/7318
Trying to fix #7318.
I fixed all issues except PY2.

kimdwkimdw committed:
Add missing `Type Hints` to torch. issue #7318

apaszke commented on torch/_torch_docs.py:
```diff
@@ -68,7 +69,7 @@ def parse_kwargs(desc):
 """)
 
-add_docstr(torch.abs,
+add_docstr(_C.abs,
```
What does this fix? torch.* and _C.* functions are exactly the same object so precious code should be fine

apaszke commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -269,18 +269,23 @@ static PyObject * THPVariable_tensor(PyObject* self, PyObject* args, PyObject* k
+  {"saddmm", (PyCFunction)THPVariable_sspaddmm, METH_VARARGS | METH_KEYWORDS, NULL},
+  {"sparse_coo_tensor", (PyCFunction)THPVariable_sparse_coo_tensor, METH_VARARGS | METH_KEYWORDS, NULL},
+  {"spmm", (PyCFunction)THPVariable_mm, METH_VARARGS | METH_KEYWORDS, NULL},
+  {"tensor", (PyCFunction)THPVariable_tensor, METH_VARARGS | METH_KEYWORDS, NULL},
```
Why did this needed to be separated?

kimdwkimdw replied:
THPVariable_from_numpy is the only method without METH_VARARGS.

Without it, it cannot be added via PyModule_AddFunctions.

apaszke commented on tools/autograd/gen_python_functions.py:
```diff
@@ -186,7 +186,7 @@ def should_bind(declaration):
     py_torch_functions = group_declarations_by_name(declarations, should_bind)
 
-    env = create_python_bindings(py_torch_functions, has_self=False)
+    env = create_python_bindings(py_torch_functions, has_self=False, is_module=True)
```
Why did this change?

kimdwkimdw replied:
It removes METH_STATIC on torch_functions.

Without METH_STATIC, those methods can be added to _C via PyModule_AddFunctions.

kimdwkimdw committed:
reversed commit `_C -> torch` add_docstr in _torch_docs.py

kimdwkimdw committed:
resolve conflict in torch/__init__.py#L82 (__all__)

kimdwkimdw commented on torch/__init__.py:
```diff
@@ -186,6 +186,7 @@ def set_default_dtype(d):
 from .tensor import Tensor
 from .storage import _StorageBase
 
+tensor = _C.tensor  # remove duplicate name, and fix to shortcut function
```
It can move to below line of `torch/__init__.py#L250`
```
for name in dir(_C._VariableFunctions):
    globals()[name] = getattr(_C._VariableFunctions, name)
```

kimdwkimdw committed:
Fix issues about flake8 in travis-ci.org

kimdwkimdw committed:
fix error with `'torch' has no attribute '_*` from _C

https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-macos-10.13-py3-test1/1965/console

kimdwkimdw committed:
fix `failure of builds` about private method in torch_functions_additional

zou3519 commented:
Thank you for the PR, @kimdwkimdw! What is the general strategy for this PR? From what I can tell it looks like you're removing METH_STATIC, but why does that affect how pycharm can read the code?

kimdwkimdw commented:
@zou3519 , Thank you for comment!

First, here is my concept of gist, 
https://gist.github.com/kimdwkimdw/50c18b5cf72c69c2d01bb4146c8a2b5c

From 0.4.0, all methods in torch_functions goes to `_C._VariableFunctions` class.
With [PyModule_AddObject](https://docs.python.org/3/c-api/module.html#c.PyModule_AddObject), all methods are instance methods. Thats why we need below lines.

```
# torch/__init__.py
for name in dir(_C._VariableFunctions):
    globals()[name] = getattr(_C._VariableFunctions, name)
```

The shorthand of above is that all methods are not imported with pythonic way.

I want to import all methods via below code.
```
# torch/__init__.py
from torch._C import *
```

https://docs.python.org/3/c-api/structures.html#METH_STATIC
With `METH_STATIC`, C Extension will generate kind of static methods. However, we don't need @staticmethod (https://docs.python.org/3/library/functions.html#staticmethod) for method forwarding. We just only need `module methods`.

When we import module methods like below,
```
# torch/__init__.py
from torch._C import *
```
IDE will run `__init__.py` internally, and generate stub files following [PEP-0484](https://www.python.org/dev/peps/pep-0484/#stub-files)

There is an important sentence.
```
Modules and variables imported into the stub are not considered exported from the stub unless the import uses the import ... as ... form or the equivalent from ... import ... as ... form.
...
..
..
Stub files may be incomplete. To make type checkers aware of this, the file can contain the following code:

def __getattr__(name) -> Any: .
```

kimdwkimdw committed:
fix tests for PY2

kimdwkimdw commented on tools/autograd/gen_python_functions.py:
```diff
@@ -186,7 +189,10 @@ def should_bind(declaration):
     py_torch_functions = group_declarations_by_name(declarations, should_bind)
 
-    env = create_python_bindings(py_torch_functions, has_self=False)
+    if PY2:
```
[`PyModule_AddFunctions`](https://docs.python.org/3/c-api/module.html#c.PyModule_AddFunctions) don't exists in PY2.
generated codes conditionally with `#if PY_MAJOR_VERSION == 2`.

zou3519 commented:
@kimdwkimdw I'm not very familiar with what we're doing with METH_STATIC and what PyModule_AddFunctions is doing. Someone (probably me) will have to sit down and read the specs for these things for a bit, I will get back to you soon after I understand this a little more

kimdwkimdw commented:
@zou3519 👍  I'll think about better solution and substitute for PY2.

zou3519 commented:
Sorry for the delay in reviewing, @kimdwkimdw. I will get to it soon

zou3519 commented:
@kimdwkimdw I wonder if making `_C._VariableFunctions` a module and then creating the torch_functions on the `_C._VariableFunctions` module will fix the problem for both Python 2 and Python 3. I'll try to look more into it, but what do you think?

kimdwkimdw commented:
@zou3519 I'm not sure. But, I don't think that can solve its problem easily. 

Originally, `pytorch 0.3.x` have all torch functions in `torch._C` with `PyModule_Create` and `Py_InitModule3`.
However, `pytorch 0.4.0` divided `torch._C` into multiple _***objects***_ like `_C._VariableFunctions`. 

I think we can think multiple solutions
1) Move all code to torch._C gracefully for Python 2 like pytorch 0.3.0.
All torch functions and other functions are added with `PyModule_AddObject`, not `Module`.

Too lots of code change are needed

2) Make multiple C Extensions and connect them. Then we can use your suggestion
```
#if PY_MAJOR_VERSION >= 3
    m = PyModule_Create(&moduledef);
#else
    m = Py_InitModule3("themodulename",
```

Also too lots of code change are needed

3) Merge it, and wait for deprecation of PY2. https://pythonclock.org/ lol.

zou3519 commented:
I tried (2), it's a little painful to get it to work.
For (3), RHEL announced support for Py2 until 2022 or 2024 so it is not going away anytime soon.

For (1), we have this problem that we're dumping all functions into torch._C. It would be nice to do some namespacing and put the variable functions into torch._C.VariableFunctions or something, where VariableFunctions is not a class.

HWiese1980 commented:
Is this going to be merged anytime soon?

sproshev commented:
Generated `/torch/_C/__init__.py` is much longer (7600 LOC vs 700) after applying this PR.

We're using https://github.com/JetBrains/intellij-community/blob/master/python/helpers/generator3.py to generate stubs for binaries. I have not investigated yet what is the difference between doing so for released pytorch and pytorch with changes in this PR applied.

![screenshot_2018-10-23_19-25-51](https://user-images.githubusercontent.com/4990763/47375692-7ef53500-d6f9-11e8-9492-04b5b7abd290.png)

niushencheng commented:
anyone can help solve this issue? I almost give up pytorch for the boken  type hints

MarkVdBergh commented:
I'm having the same type hints problem with pytorch 1.0.0
Will this solution work with that version?

t-vi commented:
I have a workshop coming up where people want to use them, so I'll refresh my PR for the typehints before early next week. I'll send a link for the PYI for 1.0 when I have it.
